channel/player.go

```diff
@@ -154,14 +161,70 @@ type player struct {
 	UpdatePartyInfo updatePartyInfoFunc
 
 	rates *rates
+
+	// Per-player RNG for deterministic randomness
+	rng *rand.Rand
+}
+
+// SeedRNGDeterministic seeds the per-player RNG using stable identifiers so
+// gain sequences are reproducible across restarts and processes.
+func (d *player) SeedRNGDeterministic() {
+	// Compose as uint64 to avoid int64 constant overflow, then cast at runtime.
+	const gamma uint64 = 0x9e3779b97f4a7c15
+	seed64 := gamma ^
+		(uint64(uint32(d.id)) << 1) ^
+		(uint64(uint32(d.accountID)) << 33) ^
+		(uint64(d.worldID) << 52)
+
+	seed := int64(seed64) // two's complement wrapping is fine for rand.Source
+	d.rng = rand.New(rand.NewSource(seed))
+}
+
+// ensureRNG guarantees d.rng is initialized. If a deterministic seed has not
+// been set yet, it will use a time-based seed (non-deterministic).
+func (d *player) ensureRNG() {
+	if d.rng == nil {
+		// Default to deterministic seeding for stability unless you want variability:
+		d.SeedRNGDeterministic()
+	}
+}
+
+func (d *player) randIntn(n int) int {
+	d.ensureRNG()
+	return d.rng.Intn(n)
+}
+
+// levelUpGains returns (hpGain, mpGain) using per-player RNG and job family.
+// The random component uses a small range similar to legacy behavior.
+// Tweak the constants to match your balance targets if needed.
+func (d *player) levelUpGains() (int16, int16) {
```

Comment:
FYI: I believe there is an int scalling component to MP gains which is what fueled "HP washing"